rp2040-hal/src/adc.rs

```diff
@@ -124,20 +134,71 @@ impl Channel<Adc> for TempSense {
     }
 }
 
-impl<WORD, PIN> OneShot<Adc, WORD, PIN> for Adc
-where
-    WORD: From<u16>,
-    PIN: Channel<Adc, ID = u8>,
-{
-    type Error = ();
+/// Analog to Digital Convertor (ADC).
+///
+/// Represents an ADC within the RP2040. Each ADC has multiple channels, and each
+/// channel is either associated with a specific GPIO pin or attached to the internal
+/// temperature sensor. You should put the relevant pin into ADC mode by creating an
+/// [`AdcPin`] object with it, or you can put the ADC into `Temperature Sensing Mode"
+/// by calling [`Adc::take_temp_sensor()`]. Either way, the resulting objects can be
+/// passed to the [`OneShot::read()`] trait method to actually do the read.
+pub struct Adc {
+    device: ADC,
+}
 
-    fn read(&mut self, _pin: &mut PIN) -> nb::Result<WORD, Self::Error> {
-        let chan = PIN::channel();
+impl Adc {
+    /// Create new adc struct and bring up adc
+    pub fn new(device: ADC, resets: &mut RESETS) -> Self {
+        device.reset_bring_down(resets);
+        device.reset_bring_up(resets);
 
-        if chan == TEMPERATURE_SENSOR_CHANNEL {
-            self.device.cs.modify(|_, w| w.ts_en().set_bit())
-        }
+        // Enable adc
+        device.cs.write(|w| w.en().set_bit());
+
+        // Wait for adc ready
+        while !device.cs.read().ready().bit_is_set() {}
+
+        Self { device }
+    }
+
+    /// Free underlying register block
+    pub fn free(self) -> ADC {
+        self.device
+    }
+
+    /// Read single
+    pub fn read_single(&self) -> u16 {
+        self.device.result.read().result().bits()
+    }
+
+    /// Enable temperature sensor, returns a channel to use.
+    ///
+    /// This can only be done once before calling [`disable_temp_sensor`]. If the sensor has already
+    /// been enabled, this method will panic.
+    pub fn enable_temp_sensor(&mut self) -> TempSense {
+        self.take_temp_sensor()
+            .expect("Temp sensor is already enabled.")
```

Comment:
I'd prefer to avoid "hidden panics", functions that panic internally, and that don't have a name making that possibility obvious.

The various '.take()' methods, which are typically called once at boot and are obviously not allowed to be called multiple times are fine.

But otherwise, I consider a panic within HAL code a smell that should be avoided.

Reply:
We could mark this method as deprecated and remove it in a future release. Do you think that's a good solution?

We may not need to worry too much about deprecation considering the amount of breaking changes anyway 🤷

Reply:
Both deprecation or just removing the function would be fine.

What I'd really like to have is some API that can't fail at all. There's no inherent reason why calling "enable_temp_sensor" twice should not be allowed. Things don't fail if someone calls `set_high()` twice on a GPIO pin either. But of course I understand the reasoning here. Allowing for multiple calls to this function, _and_ making sure that the `TempSense` token doesn't get used after the flag has been disabled, _and_ avoiding additional bookkeeping overhead. I don't know if there is a way to do all this without having a `panic` somewhere.

Reply:
I added a deprecation note.